### ISSUE TITLE
Discussion: Abstract out prop mapping to independent lib

### ISSUE THREAD

**Issue by cdaringe** [Semantic-Org/Semantic-UI-React#1601]
as usual, i must always begin w/ deep gratitude to @levithomason & @layershifter for kicking so much butt on this project. thanks gents.

# problem statement

I want to borrow/steal your code, but I'm not clear how.

# context

designing components that gracefully, happily merge user-provided props in with the default props a component provides can be tedious/verbose/difficult.  I think RSUI's answer to this problem is [this babel plugin](https://github.com/layershifter/babel-plugin-transform-react-handled-props)?  example of this discussed below.

# discussion

in each RSUI component, `render()` juggles classnames, styles, etc, and carefully drops props into each root export.  for example, see [here](https://github.com/Semantic-Org/Semantic-UI-React/blob/master/src/elements/Step/Step.js#L102).

we wrote a little utility [here](https://github.com/Tripwire/octagon/blob/master/src/util/externalAttributeFilter.js) that sniffs user `props`, and returns those which we will pass to the root node.

```js
function MyCompontent (props) {
  var rest = externalAttributeFilter(props)
  return <div {...rest}>...</div>
}
```

our solution ^above^ does _not_ gracefully handle classNames, styles, etc.  i'm interested in exploring a solution that does, and not creating a snowflake.

- is the RSUI team content with how `{...rest}` props are generated?
- are there ways to enable graceful generation of `{...rest}` properties external of this project?
- if there are areas to improve here to make ^^ possible, a) would you be interested, b) can I help!

thanks for the time.

**Comment by levithomason** [comment]
>is the RSUI team content with how {...rest} props are generated?

Very much so.

>are there ways to enable graceful generation of {...rest} properties external of this project?

The babel plugin you noted does exactly this.  We also use `lib/getUnhandledProps.js`, though, it now just partitions props based on the `handledProps` generated by the babel plugin.

>if there are areas to improve here to make ^^ possible, a) would you be interested, b) can I help!

We're definitely interested in this.  You can see thoughts on a base component that handles all these tasks here #419.  I'd like to pull all this functionality into a HOC.

**Comment by cdaringe** [comment]
rad!  should we just close this and focus _over there_?  this conversation really is a sub-discussion (i think) of #419 as this topic must be addressed transitively to complete #419.
